File: internal_testsuites/golang/test_helpers/enclave_setup.go
package test_helpers

import (
	"context"
	"fmt"
	"testing"
	"time"

	"github.com/kurtosis-tech/kurtosis/api/golang/v2/core/lib/enclaves"
	"github.com/kurtosis-tech/kurtosis/api/golang/v2/util"
	"github.com/kurtosis-tech/stacktrace"
	"github.com/sirupsen/logrus"
	"github.com/stretchr/testify/require"
)

const (
	testsuiteNameEnclaveIDFragment = "go-test"

	destroyEnclaveRetries                  = 3
	destroyEnclaveRetriesDelayMilliseconds = 1000
)

func CreateEnclave(t *testing.T, ctx context.Context, testName string) (resultEnclaveCtx *enclaves.EnclaveContext, resultStopEnclaveFunc func(), resultDestroyEnclaveFunc func() error, resultErr error) {
	enclaveName := fmt.Sprintf(
		"%v-%v-%v",
		testsuiteNameEnclaveIDFragment,
		testName,
		time.Now().Unix(),
	)
	enclaveCtx, stopEnclaveFunc, destroyEnclaveFunc, err := util.CreateEmptyEnclave(ctx, enclaveName)
	require.NoError(t, err, "An error occurred creating enclave '%v'", enclaveName)
	stopEnclaveFuncWrapped := func() {
		if err := stopEnclaveFunc(); err != nil {
			logrus.Errorf("An error occurred stopping enclave '%v' that we created for this test:\n%v", enclaveName, err)
			logrus.Errorf("ACTION REQUIRED: You'll need to stop enclave '%v' manually!!!!", enclaveName)
		}

	}
	destroyEnclaveFuncWrapped := func() error {
		for i := 0; i < destroyEnclaveRetries; i++ {
			if err := destroyEnclaveFunc(); err != nil {
				logrus.Warnf("An error occurred destroying enclave '%v' that we created for this test:\n%v", enclaveName, err)
				if i == destroyEnclaveRetries-1 {
					logrus.Errorf("An error occurred destroying enclave '%v' that we created for this test:\n%v", enclaveName, err)
					logrus.Errorf("ACTION REQUIRED: You'll need to destroy enclave '%v' manually!!!!", enclaveName)
					return stacktrace.NewError("An error occurred after trying to destroy the enclave '%v' %d times", enclaveName, destroyEnclaveRetries)
				}
				logrus.Warnf("Retrying %d more time(s)", destroyEnclaveRetries-i-1)
				time.Sleep(time.Duration(destroyEnclaveRetriesDelayMilliseconds) * time.Millisecond)
			} else {
				break
			}
		}
		return nil
	}

	return enclaveCtx, stopEnclaveFuncWrapped, destroyEnclaveFuncWrapped, nil
}

File: internal_testsuites/typescript/src/test_helpers/enclave_setup.ts
import { EnclaveContext, EnclaveUUID, KurtosisContext } from "kurtosis-sdk"
import {Result, err, ok} from "neverthrow"
import log from "loglevel";

const TEST_SUITE_NAME_ENCLAVE_UUID_FRAGMENT = "ts-testsuite";
const MILLISECONDS_IN_SECOND = 1000;

export async function createEnclave(testName:string, isPartitioningEnabled: boolean):
	Promise<Result<{
		enclaveContext: EnclaveContext,
		stopEnclaveFunction: () => void
		destroyEnclaveFunction: () => Promise<Result<null, Error>>,
	}, Error>> {

	const newKurtosisContextResult = await KurtosisContext.newKurtosisContextFromLocalEngine()
	if(newKurtosisContextResult.isErr()) {
		log.error(`An error occurred connecting to the Kurtosis engine for running test ${testName}`)
		return err(newKurtosisContextResult.error)
	}
	const kurtosisContext = newKurtosisContextResult.value;

	const enclaveId:EnclaveUUID = `${TEST_SUITE_NAME_ENCLAVE_UUID_FRAGMENT}.${testName}.${Math.round(Date.now()/MILLISECONDS_IN_SECOND)}`
	const createEnclaveResult = await kurtosisContext.createEnclave(enclaveId, isPartitioningEnabled);

	if(createEnclaveResult.isErr()) {
		log.error(`An error occurred creating enclave ${enclaveId}`)
		return err(createEnclaveResult.error)
	}

	const enclaveContext = createEnclaveResult.value;

	const stopEnclaveFunction = async ():Promise<void> => {
		const stopEnclaveResult = await kurtosisContext.stopEnclave(enclaveId)
		if(stopEnclaveResult.isErr()) {
			log.error(`An error occurred stopping enclave ${enclaveId} that we created for this test: ${stopEnclaveResult.error.message}`)
			log.error(`ACTION REQUIRED: You'll need to stop enclave ${enclaveId} manually!!!!`)
		}
	}

	const destroyEnclaveFunction = async ():Promise<Result<null, Error>> => {
		const destroyEnclaveResult = await kurtosisContext.destroyEnclave(enclaveId)
		if(destroyEnclaveResult.isErr()) {
			const errMsg = `An error occurred destroying enclave ${enclaveId} that we created for this test: ${destroyEnclaveResult.error.message}`
			log.error(errMsg)
			log.error(`ACTION REQUIRED: You'll need to destroy enclave ${enclaveId} manually!!!!`)
			return err(new Error(errMsg))
		}
		return ok(null)
	}

	return ok({ enclaveContext, stopEnclaveFunction, destroyEnclaveFunction })
}
